amulet/deployer.py

```diff
@@ -330,6 +337,53 @@ def deploy_w_timeout_and_dir(self, timeout, deploy_dir):
             yield
         self.deployed = True
 
+    def action_defined(self, service):
+        """
+        :param service: Name of service to get list of defined actions for.
+
+        Returns the list of actions defined for the service.
+        """
+        if service not in self.services:
+            raise ValueError('Service needs to be added before listing actions.')
+        raw = juju(['action', 'defined', service, '--format', 'json'])
+        return json.loads(raw)
+
+    def action_do(self, unit, action, **kw):
+        """
+        :param unit: Unit to run action on.
+        :param action: Action to run.
+        :param kw: Action parameters.
+
+        Runs specified action on specified unit and returns the uuid to fetch results by.
+        """
+        if '/' not in unit:
+            raise ValueError('%s is not a unit' % unit)
+        cmd = ['action', 'do', unit, action, '--format', 'json']
+        for key, value in kw.items():
+            cmd += [str(key), str(value)]
+        result = juju(cmd)
+        action_result = json.loads(result)
+        results_id = action_result["Action queued with id"]
+        return results_id
+
+    def action_fetch(self, action_id, wait=True, timeout=600):
+        """
+        :param action_id: Id of the action to fetch results for.
+        """
+        now = time()
+        while time() - now < timeout:
```

Comment:
You can avoid this loop and have it block by doing `juju action fetch --wait=0` which will block until the action returns a completed or failed